certora/DssLitePsm.spec

```
@@ -0,0 +1,1085 @@
+    require vatIlkDust == 0;
+    mathint daiJoinLive = daiJoin.live();
+    mathint vatDaiPsm = vat.dai(currentContract);
+    mathint vatDaiDaiJoin = vat.dai(daiJoin);
+    mathint vatCanDaiJoinPsm = vat.can(currentContract, daiJoin);
+    mathint gemBalanceOfPocket = gem.balanceOf(pocket);
+    require dai.wards(daiJoin) == 1;
+    mathint daiTotalSupply = dai.totalSupply();
+    require daiTotalSupply >= to_mathint(dai.balanceOf(currentContract));
+
+    mathint wad = min(
+                        min(
+                            subCap(gemBalanceOfPocket * to18ConversionFactor + buf, vatIlkArt),
+                            subCap(vatIlkLine / RAY(), vatIlkArt)
+                        ),
+                        subCap(vatLine, vatDebt) / RAY()
+                    );
+
+    fill@withrevert(e);
+
+    bool revert1  = e.msg.value > 0;
+    bool revert2  = vatIlkRate != RAY();
+    bool revert3  = gemBalanceOfPocket * to18ConversionFactor + buf > max_uint256;
+    bool revert4  = wad == 0;
+    bool revert5  = vatLive != 1;
+    bool revert6  = vatUrnArt + wad > max_uint256;
+    bool revert7  = vatIlkArt + wad > max_uint256;
+    bool revert8  = wad * RAY() > max_int256();
+    bool revert9  = vatDebt + (wad * RAY()) > max_uint256;
+    bool revert10 = vatUrnInk * vatIlkSpot > max_uint256;
+    bool revert11 = (vatUrnArt + wad) * RAY() > vatUrnInk * vatIlkSpot;
+    bool revert12 = vatDaiPsm + wad * RAY() > max_uint256;
+    bool revert13 = daiJoinLive != 1;
+    bool revert14 = vatCanDaiJoinPsm != 1;
+    bool revert15 = vatDaiDaiJoin + (wad * RAY()) > max_uint256;
+    bool revert16 = daiTotalSupply + wad > max_uint256;
+
+    assert revert1  => lastReverted, "revert1 failed";
+    assert revert2  => lastReverted, "revert2 failed";
+    assert revert3  => lastReverted, "revert3 failed";
+    assert revert4  => lastReverted, "revert4 failed";
+    assert revert5  => lastReverted, "revert5 failed";
+    assert revert6  => lastReverted, "revert6 failed";
+    assert revert7  => lastReverted, "revert7 failed";
+    assert revert8  => lastReverted, "revert8 failed";
+    assert revert9  => lastReverted, "revert9 failed";
+    assert revert10 => lastReverted, "revert10 failed";
+    assert revert11 => lastReverted, "revert11 failed";
+    assert revert12 => lastReverted, "revert12 failed";
+    assert revert13 => lastReverted, "revert13 failed";
+    assert revert14 => lastReverted, "revert14 failed";
+    assert revert15 => lastReverted, "revert15 failed";
+    assert revert16 => lastReverted, "revert16 failed";
+    assert lastReverted => revert1  || revert2  || revert3  ||
+                           revert4  || revert5  || revert6  ||
+                           revert7  || revert8  || revert9  ||
+                           revert10 || revert11 || revert12 ||
+                           revert13 || revert14 || revert15 ||
+                           revert16, "Revert rules are not covering all the cases";
```

Couldn't this be written as:
```suggestion
    assert lastReverted <=> revert1  || revert2  || revert3  ||
                            revert4  || revert5  || revert6  ||
                            revert7  || revert8  || revert9  ||
                            revert10 || revert11 || revert12 ||
                            revert13 || revert14 || revert15 ||
                            revert16, "Revert rules are not covering all the cases";
```

---

Doesn't seem equivalent as now it is sufficient for only one of these 16 conditions to be reverting for the left implication <= to be true (whereas before they all needed to result in a revert).

---

Right... the sequential reverts should read as `&&`.

---

Didn't know bidirectional implications were in the syntax now! I think this is actually a valid suggestion...`A \/ B => C` also implies `A => C` and `B => C` b/c `A` and `B` both imply `A \/ B`. That said it would make debugging a little harder if there's a revert condition that doesn't cause a revert because it wouldn't be singled out in an error message.

---

Ah yes good point, I got confused there.

---

> Didn't know bidirectional implications were in the syntax now! I think this is actually a valid suggestion...`A \/ B => C` also implies `A => C` and `B => C` b/c `A` and `B` both imply `A \/ B`. That said it would make debugging a little harder if there's a revert condition that doesn't cause a revert because it wouldn't be singled out in an error message.

Is it actually a real problem? Certora will still show the variables and you could identify which is/are one/s that are returning true without a revert condition.
The only negative side I can see is that putting everything in a bidirectional assert might affect speed due some lack of possible parallelization when multi-assert is enabled. However might be  worth a try to test this out as it would be a very big spec code reduction.

---

IMO it is not a huge issue, as that's not often something you need to debug, and it's still possible like you say, just slightly slower. I'd be in favor of the change if there is no significant performance degradation.

---

So I tried this and indeed I'm getting a timeout on `fill_revert` that we are not getting with the actual structure. So I'd leave this idea on hold for now.